### ISSUE TITLE
Errors after running :EnClasspath

### ISSUE THREAD

**Issue by chipsenkbeil** [ensime/ensime-vim#280]
So, I'm trying to set up vim with ensime to work on the ensime-server project. From reading the docs online, my assumption was that I do the following:

1. Run `sbt ensimeConfig` from the root of the project
2. Start vim
3. Run `:EnClasspath` once when first starting the project

After running `:EnClasspath`, I'm told to hit enter to continue after various libraries are resolved and downloaded. This results in me returning to vim and seeing this error:

```
Error detected while processing function ensime#com_en_classpath[1]..<SNR>161_call_plugin:
E605: Exception not caught: <type 'exceptions.OSError'>: [Errno 2] No such file or directory
```

Furthermore, attempting to quit (via `:q`) freezes vim, which no longer responds (I can't even interrupt or suspend it). I have to search for the process and kill it via `kill -9 <proc id>`.

Did I miss a step? Something else going wrong? My JVM is as follows:
```
openjdk version "1.8.0_92"
OpenJDK Runtime Environment (build 1.8.0_92-b14)
OpenJDK 64-Bit Server VM (build 25.92-b14, mixed mode)
```

Consecutive calls to `:EnClasspath` without quitting the vim session results in the same error being printed out (but nothing being resolved, just stays in the vim session).

**Comment by ChrisCoffey** [comment]
@chipsenkbeil is there any additional information in the `.ensime_cache/ensime-vim.log`?

**Comment by chipsenkbeil** [comment]
Removed it, opened vim and ran `:EnClasspath` again, and here is the resulting log:

```
2016-06-09 10:28:29.924735: __init__: in
2016-06-09 10:28:29.964047: [(<frame object at 0x7f8d7fb2fde0>, '/home/senkwich/.vim/bundle/ensime-vim/ensime_shared/ensime.py', 208, 'lazy_initialize_ensime', ['                self.log(str(inspect.stack()))\n'], 0), (<frame object at 0x7f8d7e9db050>, '/home/senkwich/.vim/bundle/ensime-vim/ensime_shared/ensime.py', 228, 'setup', ['        return self.running and lazy_initialize_ensime() and ready_to_connect()\n'], 0), (<frame object at 0x7f8d8178b650>, '/home/senkwich/.vim/bundle/ensime-vim/ensime_shared/ensime.py', 1182, 'client_for', ['            if client.setup(quiet=quiet, create_classpath=create_classpath):\n'], 0), (<frame object at 0x22b2730>, '/home/senkwich/.vim/bundle/ensime-vim/ensime_shared/ensime.py', 1158, 'current_client', ['                create_client=create_client)\n'], 0), (<frame object at 0x2286910>, '/home/senkwich/.vim/bundle/ensime-vim/ensime_shared/ensime.py', 1100, 'wrapper2', ['                create_client=create_client)\n'], 0), (<frame object at 0x7f8d8179f228>, '<string>', 3, '<module>', None, None)]
2016-06-09 10:28:29.964289: setup(False, True) called by wrapper2()
```

**Comment by jvican** [comment]
Which operating system?

**Comment by chipsenkbeil** [comment]
ArchLinux. And vim statistics are as follows:

```
VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Apr 29 2016 11:28:12)
Included patches: 1-1797
Compiled by Arch Linux
Huge version with GTK2 GUI.  Features included (+) or not (-):
+acl             +farsi           +mouse_netterm   +tag_binary
+arabic          +file_in_path    +mouse_sgr       +tag_old_static
+autocmd         +find_in_path    -mouse_sysmouse  -tag_any_white
+balloon_eval    +float           +mouse_urxvt     +tcl/dyn
+browse          +folding         +mouse_xterm     +terminfo
++builtin_terms  -footer          +multi_byte      +termresponse
+byte_offset     +fork()          +multi_lang      +termtruecolor
+channel         +gettext         -mzscheme        +textobjects
+cindent         -hangul_input    +netbeans_intg   +timers
+clientserver    +iconv           +packages        +title
+clipboard       +insert_expand   +path_extra      +toolbar
+cmdline_compl   +job             +perl/dyn        +user_commands
+cmdline_hist    +jumplist        +persistent_undo +vertsplit
+cmdline_info    +keymap          +postscript      +virtualedit
+comments        +langmap         +printer         +visual
+conceal         +libcall         +profile         +visualextra
+cryptv          +linebreak       +python/dyn      +viminfo
+cscope          +lispindent      +python3/dyn     +vreplace
+cursorbind      +listcmds        +quickfix        +wildignore
+cursorshape     +localmap        +reltime         +wildmenu
+dialog_con_gui  +lua/dyn         +rightleft       +windows
+diff            +menu            +ruby/dyn        +writebackup
+digraphs        +mksession       +scrollbind      +X11
+dnd             +modify_fname    +signs           -xfontset
-ebcdic          +mouse           +smartindent     +xim
+emacs_tags      +mouseshape      +startuptime     +xsmp_interact
+eval            +mouse_dec       +statusline      +xterm_clipboard
+ex_extra        +mouse_gpm       -sun_workshop    -xterm_save
+extra_search    -mouse_jsbterm   +syntax          -xpm
   system vimrc file: "/etc/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "/etc/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/usr/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread -I/usr/include/gtk-2.0 -I/usr/lib/gtk-2.0/include -I/usr/include/pango-1.0 -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/libdrm -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng16 -I/usr/include/pango-1.0 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/harfbuzz -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/harfbuzz -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -D_FORTIFY_SOURCE=2  -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1      
Linking: gcc   -L. -Wl,-O1,--sort-common,--as-needed,-z,relro -fstack-protector -rdynamic -Wl,-export-dynamic -Wl,-E -Wl,-rpath,/usr/lib/perl5/core_perl/CORE  -Wl,-O1,--sort-common,--as-needed,-z,relro -L/usr/local/lib -Wl,--as-needed -o vim   -lgtk-x11-2.0 -lgdk-x11-2.0 -lpangocairo-1.0 -latk-1.0 -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lpangoft2-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lfontconfig -lfreetype -lSM -lICE -lXt -lX11 -lXdmcp -lSM -lICE  -lm -lncurses -lelf -lnsl    -lacl -lattr -lgpm -ldl   -Wl,-E -Wl,-rpath,/usr/lib/perl5/core_perl/CORE -Wl,-O1,--sort-common,--as-needed,-z,relro -fstack-protector-strong -L/usr/local/lib  -L/usr/lib/perl5/core_perl/CORE -lperl -lpthread -lnsl -ldl -lm -lcrypt -lutil -lc   -L/usr/lib -ltclstub8.6 -ldl -lz -lpthread -lieee -lm     
```

**Comment by jvican** [comment]
Go to `~/.config/` and remove any ensime-related folder. Then, try again. Also, just in case, regenerate the ensime config and make sure you've got the latest version of the server.

**Comment by chipsenkbeil** [comment]
Noticed that I'm getting several warnings regarding `Unable to reparse org.ensime#..., using Tue May 31 16:41:11 CDT 2016` during the resolution step. So, I'm assuming that I'm stuck with the version released at that point. Still hanging after clearing the ensime-vim directory (only ensime-related directory).

**Comment by ches** [comment]
I'm suspecting either a permissions problem somewhere, or possibly your Java home isn't getting picked up correctly somehow. So bear with me on a string of possibly weird questions, I'm just tracing through the code based on the stack in your log and trying to rule things out where exceptions might possibly not be caught currently:

- You do have an `.ensime` file present in the project directory resulting from `sbt ensimeConfig`, right?
- Can you verify that the value of `:java-home` within that `.ensime` file is a valid JDK installation, i.e. `bin/java` within that path exists and is executable by your user, etc.?
- Has a file been created at `~/.config/ensime-vim/<your project Scala version>/classpath`?
- Has an `.ensime_cache/` directory been created in the project directory, and `server.log` and `server.pid` files in there?
- Is there any symlink in the path to your project? Anything interesting about where the project lives on disk, like an unusual volume?
- This is the craziest one, but it is a file read call outside of a `try`, so… does `/dev/null` exist on your system and readable to your user?

**Comment by chipsenkbeil** [comment]
@ches, sorry it took so long to get back to you on this. Work takes priority and since I haven't been able to switch to Ensime yet, I can only dabble during early hours  and off time.

Anway, I've immediately noticed that my `:java-home` within `.ensime` points to `"/home/senkwich/projects/personal/ensime-server"` as opposed to a valid JDK directory. Not sure how that happened.

There is a file located at `~/.config/ensime-vim/2.11.8/classpath`.

There is an `.ensime_cache/server.log` file that is empty (along with an `ensime-vim.log` file), but no `server.pid`. I'm assuming that gets deleted when the server isn't running, though, which is the case for me.

For this machine, I have a single partition located at `/` that contains my home directories along with everything else. Only temporary directories like `/tmp` use _tmpfs_. So, nothing usual about where the project lives, which is `/home/senkwich/projects/personal/ensime-server`.

Oh, and `/dev/null` does exist on my system. ;)

I ran `sbt ensimeConfig` again. The only warning I get is...

```
[warn] No Java sources detected in  (your ENSIME experience will not be as good as it could be.)
[warn] Some source directories do not exist and will be ignored by ENSIME.
[warn] For example: /home/senkwich/projects/personal/ensime-server/testing/simple/src/test/scala,/home/senkwich/projects/personal/ensime-server/testing/macros/src/main/scala,/home/senkwich/projects/personal/ensime-server/testing/timing/src/test/scala,/home/senkwich/projects/personal/ensime-server/testing/shapeless/src/test/scala,/home/senkwich/projects/personal/ensime-server/testing/debug/src/test/scala 
```

`:java-home` is still set incorrectly. When I manually correct it to `/usr/lib/jvm/java-7-openjdk`, I get `Ensime server has started...` in vim when opening `org.ensime.core.debug.DebugActor`. Tried `:EnDeclaration` and it seems to work, which is awesome!

So, should this be raised against the `sbt-ensime` project? Or is it `ensime-sbt`? Can never remember? I don't have `JAVA_HOME` or `JDK_HOME` set, so not sure if that project is expected to find my JDK otherwise.

If this should be targeted elsewhere, I'd still say that `ensime-vim` needs to gracefully handle that problem (if possible). At least avoiding a freeze of vim to the point where I have to open another terminal to kill it (can't `Ctrl+C` or `Ctrl+Z` at that point). Still, really happy that the `:java-home` appears to have been the only problem! The work I was doing in the server  has finished, but maybe I can take a look at improving `ensime-vim`'s debug capabilities now.

**Comment by fommil** [comment]
sbt-ensime will use the java that sbt is using, so you just need to fix your environment so that sbt is using the right version of java.

**Comment by ches** [comment]
Glad you've sorted it out! If a wonky `:java-home` value being generated in `.ensime` is reproducible, that would indeed not be `ensime-vim`'s issue.

> If this should be targeted elsewhere, I'd still say that `ensime-vim` needs to gracefully handle that problem (if possible). At least avoiding a freeze of vim to the point where I have to open another terminal to kill it (can't `Ctrl+C` or `Ctrl+Z` at that point).

I agree. This is actually two problems, the first being to give the user a better error message in this scenario. I've addressed that in a (unmerged) commit referenced above. It comes along with some larger refactoring so if that takes awhile to review then I'll try to extract just that change.

The second is the bad failure behavior of needing to `kill -9` Vim sometimes. This drives me nuts too, I haven't dug into why that happens yet but I suspect it has to do with threading/teardown logic that is hard to follow and thus likely buggy… I've got some kind of threading-related bug also arising when `ensime-vim` and another Python-based plugin that I use are loaded together, so that's going to drive me to debug this pretty soon…

**Comment by ches** [comment]
I think @chipsenkbeil's issue is resolved, and since #293 ensime-vim should now fail gracefully if for some reason it's directed to a `java` executable that can't be run.

I know I too have run into cases where ensime-vim hangs Vim to the point that a `kill -9` is needed, but I don't have a reproduction right now. Once I have one I'll open an issue to track it, if anyone else has a reproduction you're of course welcome to.
